content/tutorials/parallelization/SIMWE_parallelization.qmd

```diff
@@ -0,0 +1,804 @@
+                "erosion_median": erosion["median"],
+                "erosion_total": erosion["sum"] * region["nsres"] * region["ewres"],
+            }
+
+
+if __name__ == "__main__":
+    tools = Tools()
+    # The entire workflow will run in parallel,
+    # so this limits the number of threads the tools can use to 1.
+    tools.g_gisenv(set="NPROCS=1")
+    basins = tools.v_db_select(format="json", map="river_basins")["records"]
+    huc12s = [basin["huc12"] for basin in basins]
+    # set the number of processes:
+    with Pool(processes=cpu_count()) as pool:
+        result = list(tqdm(pool.imap(compute, huc12s), total=len(huc12s)))
+        with open("result.json", "w") as fp:
+            json.dump(result, fp)
+```
+
+Now execute the script, it will take some time.
+
+```{python}
+%run script.py
+```
+
+Load the resulting JSON file into a pandas dataframe and normalize the results, so that we can easily compare the subwatersheds.
+
+```{python}
+import pandas as pd
+import json
+
+with open("result.json") as f:
+    stats = json.load(f)
+
+df = pd.DataFrame(stats)
+df["normalized_erosion"] = df.erosion_mean / max(df.erosion_mean)
+df
+```
+
+Load the subwatershed layers into geopandas for visualization. Join the dataframe with the simulation values using the _huc12_ key and keep only subwatersheds that have computed erosion values.
+
+```{python}
+import geopandas as gpd
+
+gdf = gpd.read_file(hydro_filename, layer="WBDHU12")
+gdf = gdf.merge(df, on="huc12", how="left")
+# JSON does not support Timestamp type
+gdf = gdf.drop(columns=["loaddate"])
+gdf = gdf[gdf["erosion_total"].notna()]
+```
+
+Let's visualize the data with folium:
+
+```{python}
+import folium
+import branca.colormap as cm
+
+metric = "normalized_erosion"
+color_scale = cm.linear.OrRd_09.scale(min(gdf[metric]), max(gdf[metric]))
+color_scale.caption = "Min-max normalized mean erosion"
+# get region center as latitude, longitude
+region = tools.g_region(vector="river_basins", format="json", flags="b")
+m = folium.Map(location=[region["ll_clat"], region["ll_clon"]], zoom_start=9)
+
+tooltip = folium.GeoJsonTooltip(
+    fields=["huc12", metric],
+    aliases=["HUC12", "Normalized mean erosion:"],
+    localize=True,
+    sticky=False,
+    labels=True,
+    style="""
+        background-color: #F0EFEF;
+        border: 2px solid black;
+        border-radius: 3px;
+        box-shadow: 3px;
+    """,
+    max_width=800,
+)
+
+g = folium.GeoJson(
+    gdf,
+    style_function=lambda x: {
+        "fillColor": color_scale(x["properties"][metric]),
+        "color": "black", "weight": 1,
+        "fillOpacity": 0.5,
+    },
+    tooltip=tooltip,
+).add_to(m)
+color_scale.add_to(m)
+folium.LayerControl().add_to(m)
+m
+```
+
+![](SIMWE_images/webmap.webp)
+
+We can now identify subwatersheds that have larger erosion values and are potentially
+the source of the source of increased sediment loads downstream, contributing to water quality degradation and habitat disruption.
```

Comment:
Say this upfront even with a caveat that this will be a somewhat limited analysis.